Sequence data
=============

General informations
^^^^^^^^^^^^^^^^^^^^

Main objective
--------------
In this lecture we will introduce sequence data formats fastq and fasta. We will discuss how features such as genes are annotated in genbank and gff formats. Sequence and feature data is available via any number of online databases, and we will explore NCBI and ENA resources.


Learning objectives
-------------------

* Students can recognise and work with fastq and fasta formatted sequence files
* Students can interpret genbank and gff formatted files
* Students are able to identify an appropriate online database to find relevant sequence file
* ...

Requirements
------------

* Computer/Laptop with access to the ETHZ network, via VPN if necessary

FASTA and FASTQ files
^^^^^^^^^^^^^^^^^^^^^^
Sequence data obtained from high-throughput experiment are usually either stored in the FASTA or in the FASTQ format. Both are text files and the FASTQ file is based of the FASTA file.

FASTA
-----
FASTA is a text-based format which can represent nucleotide or protein sequences. Each file useually starts with a header whose first character is ">" and usually contains describtive information about the sequence such as the sequence ID. The next lines of the file contain the actual sequence.

.. code-block:: bash
   
   #Nucleotide sequence
   >Mus_musculus_tRNA-Ala-AGC-1-1 (chr13.trna34-AlaAGC)
   GGGGGTGTAGCTCAGTGGTAGAGCGCGTGCTTAGCATGCACGAGGcCCTGGGTTCGATCC
   CCAGCACCTCCA

   #Protein sequence
   >gi|5524211|gb|AAD44166.1| cytochrome b [Elephas maximus maximus]
   LCLYTHIGRNIYYGSYLYSETWNTGIMLLLITMATAFMGYVLPWGQMSFWGATVITNLFSAIPYIGTNLV
   EWIWGGFSVDKATLNRFFAFHFILPFTMVALAGVHLTFLHETGSNNPLGLTSDSDKIPFHPYYTIKDFLG
   LLILILLLLLLALLSPDMLGDPDNHMPADPLNTPLHIKPEWYFLFAYAILRSVPNKLGGVLALFLSIVIL
   GLMPFLHTSKHRSMMLRPLSQALFWTLTMDLLTLTWIGSQPVEYPYTIIGQMASILYFSIILAFLPIAGX
   IENY

FASTQ Format
------------
Due to how sequence techniques work and their technical limitations the confidence (the probability of having correctly identified the nucleotide) for each base call is not equal within the sequence. This quality score (usually a Phred score) could not be implemented into the FASTA format therefore the FASTQ format was introduced. Usually only sequence data is stored in FASTQ files andprotein sequences in FASTA files.


To understand the fastq format on the computer, we can copy and look at an example file.

.. code-block:: bash

    # Copy the example file to my home directory
    I need a file

    # Take a look
    less file

Each sequence consists of four lines:

1. The sequence identifier, almost certainly automatically generated by the sequencing device; the line always begins with "@"
2. The raw sequence, consisting of A, C, G, T and perhaps N
3. These days unused, but in theory a repeat of line 1; the line always begins with "+"
4. The quality scores for each base in the raw sequence

.. code-block:: bash
   
   @071112_SLXA-EAS1_s_7:5:1:817:345
   GGGTGATGGCCGCTGCCGATGGCGTC
   AAATCCCACC
   +
   IIIIIIIIIIIIIIIIIIIIIIIIII
   IIII9IG9IC

Quality checking
^^^^^^^^^^^^^^^^

Phred Quality Scores
--------------------

There were several competing systems for encoding the quality scores but the Sanger format encoding Phred quality scores has won out. This is based on the decimal code for the character used from the ASCII encoding table. For instance, the letter **A** has a decimal equivalent of **65**. Sanger format offsets the quality score by +33, so this represents a quality score of **32** for the letter **A**.

.. figure:: images/Ascii_Table-nocolor.svg
    :align: center

The Phred quality score (Q) is logarithmically related to the error probability (E) and can therefore be interpreted as an estimate of error (E) or as an estimate of accuracy (A) which is used in the FASTQ format.

.. math::
   
   & Q = -10log(E) \\
   & E = 10^{-(Q/10)} \\
   & A = 1-E = 1 - 10^{-(Q/10)}


These formulas lead to this table on how the Phred scores can be interpreted. 

+------------------+-----------+------------+----------------------+
| Phred Quality Score          |   Error    |  Accuracy            |
|                              |            |                      |
+==============================+============+======================+
| 10                           | 10%        | 90%                  |
+------------------------------+------------+----------------------+
| 20                           | 1%         | 99%                  |
+------------------------------+------------+----------------------+
| 30                           | 0.1%       | 99.9%                |
+------------------------------+------------+----------------------+
| 40                           | 0.01%      | 99.99%               |
+------------------------------+------------+----------------------+


Phred quality scores usually range from 0-40 in FASTQ files. For most purposes a Phred score of 20 is the acceptable limit of quality meaning that the base was predicted with 99% accuracy.

Quality checking on sequence reads
----------------------------------
**Explain how to check the quality scores proberly for exmaples**


Exercises
---------
* Open file X.fasta

.. hidden-code-block:: bash

   #Use nano
   nano X.fasta


* How many reads are in this file?

.. hidden-code-block:: bash

   #Each read starts with a ">" so we can count the number of ">" appearances
   grep -c ">" X.fasta


* Open the X.fastq file

.. hidden-code-block:: bash

   #Use nano
   nano X.fastq

* How many reads are in this file?

.. hidden-code-block:: bash

   #Each read starts with a "@" so we can count the number of "@" appearances
   grep -c "@" X.fastq

* Looking at the first read which bases would you exclude?
* ...


Sequence databases
^^^^^^^^^^^^^^^^^^
NCBI
----
The National Center for Biotechnology Information (NCBI) houses a series of databases and tools relevant to biotechnology, biomedicine and bioinformatics and is probably the most prominent online database in Biology. The NCBI includes major databases such as GenBank or PubMed (a literature database which you probably will use a lot in your scientific career). 

Entrez
++++++

...

RefSeq
++++++

...

GenBank
+++++++
GeneBank is a genetic sequence database which annotates and collects all publicly available DNA sequences and their protein translations.Currently, the GenBank harbours more than 12 trillion nucleotides in more than 2 billion sequences from more than 100’000 distinct organisms. The sequences are submitted from laboratories all around the world.

Genome
++++++

...

ENA
---
The European Nucleotide Archive (ENA) repository for annotated DNA and RNA sequences with complementary information. The ENA consists of three main databases: The Sequence Read Archive (SRA), the Trace Archive (TA) and the EMBL Nucleotide Archive (EMBL-bank) 

Sequence Read Archive
+++++++++++++++++++++

The SRA is a bioinformatic database maintained by the NCBI, European Bioinformatics Institute (EBI) and DNA Data Bank of Japan (DDBJ) providing access to DNA sequences, especially short reads (less than 1’000 base pairs). In the SRA primary data from next-generation-sequencing are stored but also data from RNA-Seq and ChIP-Seq.

Trace Archive
+++++++++++++
Same as for the SRA also the TA is maintained by the NCBI, EBI and DDBJ together. The TA is a permanent repository of DNA sequence chromatograms (traces), base calls and quality estimates for single-pass reads from large-scale sequencing projects. 


EMBL-bank
+++++++++

The EMBL-bank contains high high-level genome assembly details, as well as assembled sequences and their functional annotation.

DDBJ
----
The DNA Data Bank of Japan (DDBJ) is a biological database collecting DNA sequences and is the only nucleotide database in Asia and collects mainly DNA data from Japanese researchers. 


Exercise
--------

**How to acces and work with these Databases probabley NCBI**

Protein databases
^^^^^^^^^^^^^^^^^

Uniprot
-------

The Universal Protein Resource (UniProt) is a database hosted by the European Bioinformatics Institute (EMBL-EBI), Swiss Institute of Bioinformatics (SIB) and  Protein Information Resource (PIR) for protein sequence, annotation and functions. Uniprot consists of three databases the UniProt Knowledgebase (UniProtKB), the UniProt Reference Clusters (UniRef), and the UniProt Archive (UniParc). 

.. image:: images/overview.png
   

The UniProtKB consists of two sections, a manually annotated part and an automated annotated part which awaits manual annotation. Each entry provides functional information about a protein with as much annotation information as possible.

NCBI
----

...


Homework
^^^^^^^^
