Practical training in bioinformatics FS21
=========================================

Welcome to the course “GL Biology II FS2022- Bioinformatics 2021”. With the recent development in sequencing, bioinformatics became more and more important and it is crucial for a biologist to be able to handle a large amount of data correctly. This course aims to equip undergraduate students in Biology with basic competencies in performing bioinformatic analyses. The course will introduce the use of the command-line interface (UNIX/LINUX) and reinforce programming skills in python and R, building on proficiencies that were acquired during the 2nd and 3rd-semester courses “Grundlagen der Informatik” and “Statistik II”. Existing knowledge will be augmented by programmatic, analytical, and statistical principles as they pertain to biological data to support concepts taught in concurrent courses, and to prepare students for the 3rd year of the study curriculum.

Requirements
------------
**Important:** In order to participate in the course you need to fulfill the following requirements:

* You have a desktop or laptop computer with browser access and a keyboard (a decent amount of typing is involved in this course).
* Your computer or laptop has to be at ETH or connected to the Virtual Private Network (VPN) of ETH Zürich. How to set up a VPN connection is described `here <https://ethz.ch/content/dam/ethz/special-interest/hest/isg-hest-dam/documents/pdf/vpn-de.pdf>`_
* For Windows users: Windows 10 or later versions are suitable. You will need to install MobaXterm before the course
* For Mac OS X users: You will need the app “Terminal”. It is located in the “Applications” folder within the “Utilities” sub-folder. It is recommended to add it to your Dock for the duration of the course.

Instructors
-----------

|pic1|  |pic2|

Shinichi Sunagawa |nbspc| |nbspc| |nbspc| |nbspc| |nbspc| |nbspc| |nbspc| |nbspc| |nbspc| |nbspc| Chris Field

.. |nbspc| unicode:: U+00A0 .. non-breaking space

.. |pic1| image:: https://micro.biol.ethz.ch/research/sunagawa/_jcr_content/rightpar_top/contextinfo/fullwidthimage/image.imageformat.context.874464339.png
      :width: 25%

.. |pic2| image:: https://biol.ethz.ch/en/the-department/people/person-detail.person_image.jpeg?persid=MjQ1NDA0
      :width: 22%


Schedule
--------

The course will have 13 2h lectures each week on **Thursday** from **X-Y** from **24.02.2022-02.06.2022** (no course on the 26.5.2022, Auffahrt). 

The course will be held in person or via zoom (TBD).

Learning objectives
-------------------

Intro to programmig / sequence-based bioinformatics:

* Students can navigate the UNIX environment, execute basic commands and work on high-performance computing (HPC) clusters.
* Students are confident working in Python and R and able to install new packages to assist in data analysis.
* Students are familiar with standard file formats for sequences and annotations.
* Students are able to align nucleotide and protein sequences with BLAST via the web and command line and can interpret the results.
* Students are able to perform multiple alignments and reconstruct a phylogenetic tree to investigate the evolution of the whole genome and gene sequences.
* Students are able to identify proteins via Hidden Markov Models (HMMs), and can visualize and predict protein structures based on the sequence.
* Students can identify non-coding features of genomes by motif-searching and the use of Position Weight Matrices (PWMs)
* Students can explore experiment count data, such as produced by RNA-Seq, via ordination and can interpret results in the context of gene ontology term enrichment.
* Students can identify and track features in experiment images to quantify phenotypes.

Structural bioinformatics:

* Resources: Students know the most relevant databases of protein structures (PDB)
* Sequence to structure: Students are able to predict a protein structure when no structure already exists.
* Comparing structures: Students will be able to compare two structural models of the same protein and evaluate their similarities.
* The impact of mutations: Students will know how to use a protein structure to predict the impact of a mutation on protein stability.
* Protein interface regions: Students will learn how to predict potential protein-protein binding interfaces.
* Protein interactions: Students can predict the likely structure of a protein-protein or protein-molecule interaction.
* Relating sequence variation with binding: Students will be able to combine information on sequence variation with protein structures to identify mutations that could interfere with binding.

Imaging bioinformatics:

Exam
----

In order to pass the course, active participation and completion of the homework are required. The course is pass/fail and does not have a grading scale.

Material
--------

Week 1
Week 2
etc.
